src/fsharp/IlxGen.fs

```diff
@@ -2515,16 +2518,57 @@ and GenAllocUnionCase cenv cgbuf eenv (c,tyargs,args,m) sequel =
     GenAllocUnionCaseCore cenv cgbuf eenv (c,tyargs,args.Length,m)
     GenSequel cenv eenv.cloc cgbuf sequel
 
-and GenLinearExpr cenv cgbuf eenv expr sequel (contf: FakeUnit -> FakeUnit) =
+and GenLinearExpr cenv cgbuf eenv sp expr sequel (contf: FakeUnit -> FakeUnit) =
     match expr with 
     | LinearOpExpr (TOp.UnionCase c, tyargs, argsFront, argLast, m) -> 
         GenExprs cenv cgbuf eenv argsFront
-        GenLinearExpr cenv cgbuf eenv argLast Continue (contf << (fun Fake -> 
+        GenLinearExpr cenv cgbuf eenv sp argLast Continue (contf << (fun Fake -> 
             GenAllocUnionCaseCore cenv cgbuf eenv (c, tyargs, argsFront.Length + 1, m)
             GenSequel cenv eenv.cloc cgbuf sequel
             Fake))
+
+    | Expr.Sequential (e1, e2, specialSeqFlag, spSeq, _) ->
+        // Compiler generated sequential executions result in suppressions of sequence points on both
+        // left and right of the sequence
+        let spAction, spExpr =
+            (match spSeq with
+             | SequencePointsAtSeq -> SPAlways, SPAlways
+             | SuppressSequencePointOnExprOfSequential -> SPSuppress, sp
+             | SuppressSequencePointOnStmtOfSequential -> sp, SPSuppress)
+        match specialSeqFlag with
+        | NormalSeq ->
+            GenExpr cenv cgbuf eenv spAction e1 discard
+            GenLinearExpr cenv cgbuf eenv spExpr e2 sequel contf
```

Comment:
This isn't quite right the right application of the `contf` technique and we need to look very carefully at this change.  

1.  `contf` normally needs to be updated on each recursive call to `GenLinearExpr`.  I'm looking closely to work out if that's the case here.

2. The introduction of the `sp` argument is suspicious to me. In the `LinearOpExpr` case you are now pushing that `sp` into the recursive case for `argLast` which smells incorrect.

3. `contf` needs to be called on the `ThenDoSeq` branch of the code

4. the change from `SPSuppress` to `sp` in the final branch feels odd.  It might be correct but we wouldn't normally expect this change when applying the `contf` technique

Reply:
Yes, we have to be really careful here.  Changing the recursion from `GenExpr --> GenExpr` to `GenLinearExpr --> GenLinearExpr` is only valid if they do exactly the same thing.  For example the `GenExpr` code has
```
  let expr = stripExpr expr

  if not (FirstEmittedCodeWillBeSequencePoint g sp expr) then
      if EmitSequencePointForWholeExpr g sp expr then
          CG.EmitSeqPoint cgbuf (RangeOfSequencePointForWholeExpr g expr)
      elif EmitHiddenCodeMarkerForWholeExpr g sp expr then
          cgbuf.EmitStartOfHiddenCode()
  match (if compileSequenceExpressions then LowerCallsAndSeqs.LowerSeqExpr g cenv.amap expr else None) with
  | Some info ->
      GenSequenceExpr cenv cgbuf eenv info sequel
```
which is all executed each time around and the `GenLinearExpr` code doesn't.  We have to make sure they both have this logic to ensure equivalence.

@TIHan Can we pair program this from scratch?  I'd like to show you how I normally apply this `contf`  transformation  It's very close to what you have but normally when applying the approach I do it very syntactically to ensure no changes or surprises.

Reply:
Agree with all of this. You are right we need to be careful; we need to make sure we emit the sequence points correctly when we do this.

Reply:
> contf normally needs to be updated on each recursive call to GenLinearExpr.

When I first did this, I updated `contf` to compose with `GenExpr`. This is actually a problem because as soon as `GenLinearExpr` tries to gen a non-linear expr, it will call `GenExpr`. This means `GenExpr` will get called twice.